### ISSUE TITLE
System.InvalidOperationException: Collection was modified after the enumerator was instantiated.

### ISSUE THREAD

**Issue by benaadams** [xunit/xunit#1855]
Seems in Disposal
```
Unhandled Exception: System.InvalidOperationException: Collection was modified after the enumerator was instantiated.
   at System.Collections.Generic.Stack`1.Enumerator.MoveNext()
   at Xunit.Sdk.DisposalTracker.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\DisposalTracker.cs:line 26
   at Xunit.Sdk.ExtensibilityPointFactory.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\ExtensibilityPointFactory.cs:line 54
   at Xunit.Sdk.TestFramework.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestFramework.cs:line 49
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

Seen in CI https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_checked_osx10.12_innerloop_tst_prtest/7197/console

https://github.com/dotnet/coreclr/pull/21076

**Comment by bradwilson** [comment]
I suppose it's too much to hope for a simple repro?

**Comment by benaadams** [comment]
Might be a secondary error, their OSX environment seems a bit flaky

**Comment by am11** [comment]
It is one of the hard to repro issue. We seldom see this exception under high load in different legs of dotnet/runtime CI: https://github.com/dotnet/runtime/issues/11063. If we switch from `foreach` to `for` loop, I think it will prevent it from happening.

@bradwilson, if this is an acceptable change: https://github.com/am11/xunit/commit/7ecc5dcbef5a16cedc734c81b899815019ad0aa7, and issue can falls under `help wanted` category, I can issue a PR. :)

**Comment by bradwilson** [comment]
I did recently [add a bunch of double disposal protection](https://github.com/xunit/xunit/commit/9e8d5d13fbda4224de7f996d6812855650980c1c) including in `DisposalTracker`, which should stop, and in the process I stopped reusing the `DisposalTracker` from `ExtensibilityPointFactory`, which I hope will put this bug to rest.

@am11 Your fix looks like it would probably not dispose everything, because as you pop items off the stack the item count would go down, so you'll end up only disposing about the first half of the items (total count goes down as your item count goes up).

Thinking about your fix, it occurred to me that I could add even more protection here, so I did that: https://github.com/xunit/xunit/commit/eacf43523a86a907574d085c51dbcd3944c02ff3

**Comment by bradwilson** [comment]
At the very least, it should track down where the bad usage is coming from. :wink:

**Comment by am11** [comment]
Thanks for the quick fix. We will find out with the next update. 😁

> only disposing about the first half of the items

True, loop should have been inversed with the stack `for (int i = toDispose.Count; i >= 0; --i)`

I was thinking perhaps `ExtensibilityPointFactory` can be made non-static singleton and DisposalTracker can make use of `ConcurrentStack`? It might help keeping the implementation lock free.

**Comment by danmoseley** [comment]
@bradwilson would you reconsider taking this?

This continues to plague our test runs in the .NET repos, as you can see from this issue, which has just a few of the hits linked to it:
https://github.com/dotnet/runtime/issues/11063

cc @ViktorHofer

**Comment by danmoseley** [comment]
An engineer on my team is going to work on fixing this.

**Comment by maryamariyan** [comment]
> At the very least, it should track down where the bad usage is coming from. 😉

@ericstj and I digged deeper into the failures and found that for all the failures linked in https://github.com/dotnet/runtime/issues/11063 for dotnet/runtime, the issue happens on concurrent calls to `DisposalTracker.Dispose`. For these use cases, every time a test run gets finished, `TestFramework.Dispose` gets called and while nothing gets added to the stack, the issue arrives when one thread calls `Dispose` on the stack and another calls `Clear`:

https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.execution/Sdk/DisposalTracker.cs#L24-L30

Therefore, to fix the reliability issue in dotnet/runtime it should be sufficient and safest to just keep DisposalTracker as-is, but rather make sure to put a lock around access to DisposalTracker in `ExtensibilityPointFactory` only:

https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.execution/Sdk/ExtensibilityPointFactory.cs#L12-L15

**Comment by bradwilson** [comment]
Fixed by #2382
